runtime/src/local_runqueue_mts.c

```diff
@@ -0,0 +1,281 @@
+
+	uint64_t prev_pwm_deadline = priority_queue_peek(pwm->sandboxes);
+
+	/* Add the sandbox to the per-worker-module (pwm) queue */
+	int rc = priority_queue_enqueue_nolock(pwm->sandboxes, sandbox);
+	if (rc == -ENOSPC) panic("Per Worker Module queue is full!\n");
+	// debuglog("Added a sandbox to the PWM of module '%s'", sandbox->module->name);
+
+	if (priority_queue_length_nolock(pwm->sandboxes) == 1) {
+		pwm->mt_class = MT_DEFAULT;
+		if (sandbox->module->remaining_budget > 0) {
+			pwm->mt_class     = MT_GUARANTEED;
+			destination_queue = local_runqueue_mts_guaranteed;
+		}
+
+		/* Add sandbox module to the worker's timeout queue if guaranteed tenant and only if first sandbox*/
+		if (sandbox->module->replenishment_period > 0) {
+			struct module_timeout *mt = malloc(sizeof(struct module_timeout));
+
+			mt->timeout = get_next_timeout_of_module(sandbox->module->replenishment_period, __getcycles());
+			mt->module  = sandbox->module;
+			priority_queue_enqueue_nolock(worker_thread_timeout_queue, mt);
+		}
+
+		rc = priority_queue_enqueue_nolock(destination_queue, pwm);
+
+		return;
+	}
+
+	if (sandbox->absolute_deadline < prev_pwm_deadline) {
+		/* Maintain the minheap structure by Deleting and
+		 *  adding the pwm from and to the worker's local runqueue.
+		 * Do this only when the pwm's priority is updated. */
+		if (pwm->mt_class == MT_GUARANTEED) destination_queue = local_runqueue_mts_guaranteed;
+
+		rc = priority_queue_delete_nolock(destination_queue, pwm);
+		assert(rc == 0);
+
+		rc = priority_queue_enqueue_nolock(destination_queue, pwm);
+		if (rc == -ENOSPC) panic("Worker Local Runqueue is full!\n");
+
+		// debuglog("Added the PWM back to the Worker Local runqueue - %s to Heapify", QUEUE_NAME);
+	}
+}
+
+/**
+ * Deletes a sandbox from the runqueue
+ * @param sandbox to delete
+ */
+static void
+local_runqueue_mts_delete(struct sandbox *sandbox)
+{
+	assert(sandbox != NULL);
+	struct perworker_module_sandbox_queue *pwm = &sandbox->module->pwm_sandboxes[worker_thread_idx];
+
+	/* Delete the sandbox from the corresponding Per-Worker-Module queue */
+	if (priority_queue_delete_nolock(pwm->sandboxes, sandbox) == -1)
+		panic("Tried to delete sandbox %lu from PWM queue, but was not present\n", sandbox->id);
+
+	struct priority_queue *destination_queue = local_runqueue_mts_default;
+	/* Because of the race, the following if clause does not work properly, so just trying both queues */
+	/*
+	if (sandbox->module->remaining_budget > 0) {
+	        destination_queue = local_runqueue_mts_guaranteed;
+	}
+	*/
+
+	/* Delete the PWM from the local runqueue completely if pwm is empty, re-add otherwise to heapify */
+	if (priority_queue_delete_nolock(destination_queue, pwm) == -1) {
+		// TODO: Apply the composite way of PQ deletion O(logn)
+		destination_queue = local_runqueue_mts_guaranteed;
+
+		if (priority_queue_delete_nolock(destination_queue, pwm) == -1)
```

Comment:
I might be missing that the standard allows multi-line conditionals without `{}`, but I really find them dangerous.